academy/2-cosmos-concepts/11-testing.md

```diff
@@ -0,0 +1,88 @@
+
+</HighlightBox>
+
+# Testing
+
+Testing is an integral part of software development in general. Done well, it helps teams catch bugs earlier, and thus cheaper. In the context of blockchains, which can lock large amounts of value and where [migrations](./12-migrations.md) are not simply one redeploy away from a fix, bugs can be catastrophic.
+
+The Cosmos SDK implements its own testing vision for its own modules, and it would be good if your project followed the same patterns. Following the same patterns will help everyone in the ecosystem to _speak the same language_. Speaking the same language will also benefit you when you open the code for bug bounty. Indeed readable tests increase the trust that casual observers have in your code overall, and by extension, your project, and allows interested bounty hunters to be _on-boarded_ faster.
+
+## Testing pyramid
+
+After having had some [reflection](https://docs.cosmos.network/main/architecture/adr-059-test-scopes.html), Cosmos divides their tests into 4 broad categories in somewhat increasing scope:
+
+* [Unit tests](https://docs.cosmos.network/main/building-modules/testing#unit-tests)
+* [Integration tests](https://docs.cosmos.network/main/building-modules/testing#integration-tests)
+* [Simulation tests](https://docs.cosmos.network/main/building-modules/testing#simulations)
+* [End-to-end tests](https://docs.cosmos.network/main/building-modules/testing#end-to-end-tests)
+
+### Unit tests
+
+Unit tests are your classic smallest tests possible and they focus on a single module at a time. If a tested call needs something from another module, like a keeper, then:
+
+* This dependency should be [mocked](https://devopedia.org/mock-testing), including mocked responses when applicable.
+* The mock should confirm that the dependency was called as expected.
+* If applicable, the test should confirm that a mocked response was handled as expected by the module under test.
+
+To be considered as well-tested, your unit tests should cover all your module's functions.
+
+As an example, imagine that your module is moving tokens on behalf of your users. Your module therefore has a dependency on the bank keeper. As part of a unit test setup, you create a mocked bank keeper and use it. After the test action, at verification time, your unit test confirms that your module called the expected functions of the bank keeper with the expected parameters. Your unit test does **not** test whether bank balances have changed because, remember, your module does not have a real bank keeper.
+
+### Integration tests
+
+Integration tests are one step wider in scope. They are still focused on your module, but instead of mocking the dependencies, your test provisions a minimally-viable application that includes fully-fledged dependencies including, crucially, those your own module needs.
+
+A minimally-viable application is one that contains your module and all its dependencies, and their own, but nothing more.
```

Comment:
Yes, it is (1).
Yes, it continues until all dependencies are satisfied.
This may mean that in an app made with "spaghetti code", you pull everything and there are no differences between an MVA and a full app...

Reply:
Thanks!